src/userobjects/ParticlesPerElementInitializer.C

```diff
@@ -0,0 +1,284 @@
+  return params;
+}
+
+ParticlesPerElementInitializer::ParticlesPerElementInitializer(const InputParameters & parameters)
+  : InitializerBase(parameters),
+    _number_density(getParam<Real>("number_density")),
+    _particles_per_element(getParam<unsigned int>("particles_per_element")),
+    _distribution_names(getParam<std::vector<DistributionName>>("velocity_distributions"))
+{
+  if (_particles_per_element <= 0)
+    paramError("particles_per_element", "The number of particles in each element must be >= 0");
+
+  if (_number_density == 0.0)
+    paramError("number_density", "The requested number density cannot be zero.");
+}
+
+void
+ParticlesPerElementInitializer::initialSetup()
+{
+  for (const DistributionName & name : _distribution_names)
+    _velocity_distributions.push_back(&getDistributionByName(name));
+
+  if (_velocity_distributions.size() != 3)
+    paramError("velocity_distributions",
+               "You must provide 3 distributions, one for each velocity component.");
+}
+
+std::vector<InitialParticleData>
+ParticlesPerElementInitializer::getParticleData() const
+{
+
+  // counting the number of elements this process is responsible for
+  // this will allow us to allocated data structures of the appropriate length
+  unsigned int num_local_elements = 0;
+  for ([[maybe_unused]] const auto elem : *_fe_problem.mesh().getActiveLocalElementRange())
+  {
+    num_local_elements++;
+  }
+
+  // in case there are no elements for this processor do nothing
+  if (num_local_elements == 0)
+    return {};
+
+  std::vector<InitialParticleData> data =
+      std::vector<InitialParticleData>(num_local_elements * _particles_per_element);
+
+  // setting up this to be able to map from reference elemnts to the physical elemnts
+  ArbitraryQuadrature arbitrary_qrule = ArbitraryQuadrature(_mesh_dimension, FIRST);
+  FEType fe_type = FEType(CONSTANT, MONOMIAL);
+  UniquePtr<FEBase> fe = FEBase::build(_mesh_dimension, fe_type);
+  fe->attach_quadrature_rule(&arbitrary_qrule);
+  fe->get_xyz();
+  // random number generator to be reseeded on each element
+  // this only enables parallel consistency if the element ids are consistent across processes
+  MooseRandom generator;
+
+  // so long as p is a point sampled from a cube with dimensions
+  // [-1, 1] x [-1, 1] x [-1, 1]
+  // this lambda will put the point inside of the libmesh reference pyramid
+  auto put_particle_in_pyramid = [](Point & p)
+  {
+    // if the point is not in the pyramids along the x axis
+    // we need to find which one it is is in
+    if (!((p(0) - std::abs(p(2)) < 0.0) && (-p(0) - std::abs(p(2)) < 0.0) &&
+          (p(1) - std::abs(p(2)) < 0.0) && (-p(1) - std::abs(p(2)) < 0.0)))
+    {
+      // checking out the pyramids with the
+      // the x axis going through the center
+      if (((p(0) - p(1) > 0.0) && (-p(0) - p(1) > 0.0)) ||
+          ((p(0) + p(1) > 0.0) && (-p(0) + p(1) > 0.0)))
+      {
+        auto temp = p(1);
```

Comment:
```suggestion
        auto temporary = p(1);
```
To avoid confusion with `temperature`. I know there are no temperature here, but this is a pet peeve of mine...